src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs

```diff
@@ -333,33 +335,104 @@ private IList<string> GetConfigFilePaths(ISettings settings)
+               .TargetFrameworks
+               .SelectMany(f => GetPackageReferences(f.Dependencies, f.FrameworkName, assetsPackageSpec, targets))
+               .GroupBy(p => p.PackageIdentity)
+               .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
+               .ToArray();
         }
 
-        private static IEnumerable<PackageReference> GetPackageReferences(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework)
+        private IEnumerable<PackageReference> GetPackageReferences(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework, PackageSpec assetsPackageSpec = null, IList<LockFileTarget> targets = null)
         {
+            TargetFrameworkInformation assetsTargetFrameworkInformation = assetsPackageSpec?.TargetFrameworks.First(t => t.FrameworkName == targetFramework);
+
+            // If we have something from the assetsPS it means there was a change so we need to update forzosamente
+            // Check if the allowed version is the same, if it is return what we have cached
+            // else, calculate the new isntalled version
+
             return libraries
                 .Where(l => l.LibraryRange.TypeConstraint == LibraryDependencyTarget.Package)
-                .Select(l => ToPackageReference(l, targetFramework));
+                .Select(l => new BuildIntegratedPackageReference(l, targetFramework, GetVersionFromPackageSpec(l, targetFramework, assetsTargetFrameworkInformation, targets)));
+        }
+
+        private PackageIdentity GetVersionFromPackageSpec(LibraryDependency l, NuGetFramework targetFramework, TargetFrameworkInformation assetsTargetFrameworkInformation, IList<LockFileTarget> targets)
+        {
+            Utility.InstalledPackageCache installedVersion = default;
+            NuGetVersion resolvedVersion = default;
+
+            if (_installedPackages.Any())
+            {
+                if (_installedPackages.ContainsKey(targetFramework))
+                {
+                    installedVersion = _installedPackages[targetFramework].FirstOrDefault(p => p.InstalledPackage.Id.Equals(l.Name, StringComparison.OrdinalIgnoreCase));
+                    if (installedVersion != null && installedVersion.AllowedVersions.Equals(l.LibraryRange.VersionRange) && targets == null)
+                    {
+                        return installedVersion.InstalledPackage;
+                    }
+                    else
+                    {
+                        _installedPackages[targetFramework].Remove(installedVersion);
+                    }
+                }
+            }
+
+            resolvedVersion = GetInstalledVersion(l, targetFramework, assetsTargetFrameworkInformation, targets);
+
+            if (resolvedVersion == null)
+            {
+                resolvedVersion = l.LibraryRange?.VersionRange?.MinVersion ?? new NuGetVersion(0, 0, 0);
+            }
+
+            var aux = new List<Utility.InstalledPackageCache>();
```

Comment:
We could probably use a better name than aux. Same here, don't qualify the namespace for InstalledPackageCache.